src/hotspot/share/runtime/continuation.cpp

```diff
@@ -1557,34 +1557,43 @@ void FreezeBase::patch(const frame& f, frame& hf, const frame& caller, bool bott
 #endif
 }
 
-NOINLINE freeze_result FreezeBase::recurse_freeze_interpreted_frame(frame& f, frame& caller, int callee_argsize, bool callee_interpreted) {
+#ifdef ASSERT
+static void verify_frame_top(const frame& f, intptr_t* top) {
+  ResourceMark rm;
+  InterpreterOopMap mask;
+  f.interpreted_frame_oop_map(&mask);
+  assert(top <= ContinuationHelper::InterpretedFrame::frame_top(f, &mask),
+         "vsp: " INTPTR_FORMAT " Interpreted::frame_top: " INTPTR_FORMAT,
+           p2i(top), p2i(ContinuationHelper::InterpretedFrame::frame_top(f, &mask)));
+}
+#endif // ASSERT
+
+NOINLINE freeze_result FreezeBase::recurse_freeze_interpreted_frame(frame& f, frame& caller,
+                                                                    int callee_argsize,
+                                                                    bool callee_interpreted) {
 #if (defined(X86) || defined(AARCH64)) && !defined(ZERO)
   { // TODO PD
+    // yes.
+    // ContinuationHelper::InterpretedFrame::adjust_unextended_sp(f)  ?? but why?
     assert((f.at(frame::interpreter_frame_last_sp_offset) != 0) || (f.unextended_sp() == f.sp()), "");
     intptr_t* real_unextended_sp = (intptr_t*)f.at(frame::interpreter_frame_last_sp_offset);
     if (real_unextended_sp != nullptr) {
       f.set_unextended_sp(real_unextended_sp); // can be null at a safepoint
+                 // is this ever at safepoint? maybe was for preempt?
     }
   }
 #else
   Unimplemented();
 #endif
 
+  // Why 'vsp' what is 'v' ?  isn't this the sp on the real stack?
```

Comment:
To distinguish between stacks, John called the continuation stack a "horizontal stack", hence `h`, and the thread stack a "vertical stack`, hence `v`.

Reply:
Ok I wouldn't have guessed.

Reply:
s/vsp/top_sp/ or top_thread_sp ?
s/hsp/frame_sp/ ?

Reply:
s/vsp/frame_sp/
s/hsfp/heap_sp/

frame_sp can imply and looks to the reader like the normal sp in the thread stack frame.